### PR TITLE
[WIP] Library: Add PyroF4OSD hwdef

### PR DESCRIPTION
**Pull request by tenten8401**
Draft PR for the [PyroDrone F4 OSD](https://fpv-flightclub.com/products/pyrodrone-f4-osd-flight-controller/) flight controller.

@andyp1per

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+define CH_CFG_ST_RESOLUTION 16
+
+FLASH_SIZE_KB 1024
+
+
+# only one I2C bus
+I2C_ORDER I2C1
+
+# order of UARTs (and USB)
+SERIAL_ORDER OTG1 USART1 UART2 USART3 UART4 UART5
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat:23]: You have specified UART2 here ....

File: libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+# USART1
+PA9 USART1_TX USART1
+PA10 USART1_RX USART1
+
+# RC input using timer
+PA3 TIM9_CH2 TIM9 RCININT PULLDOWN
+
+# alternative RC input using UART
+# TODO: Compile error because of this?
+PA2 USART2_TX USART2 NODMA
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat:75]: ... but USART2 here (they are different)

File: libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat [hunk 3]
```diff
@@ -0,0 +1,153 @@
+# hw definition file for PyroDrone F4 OSD
+# WORK IN PROGRESS (NOT TESTED)
+
+# MCU class and specific type
+MCU STM32F4xx STM32F405xx
+
+# board ID for firmware load
+APJ_BOARD_ID 125
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat:8]: Need a different board id. There is a list somewhere of the current ones. Pick the next new one.

File: libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+define HAL_GPIO_A_LED_PIN 0
+
+# buzzer
+PB5 BUZZER OUTPUT GPIO(80) LOW
+define HAL_BUZZER_PIN 80
+define HAL_BUZZER_ON 1
+define HAL_BUZZER_OFF 0
+
+# spi1 bus for IMU
+PA5 SPI1_SCK SPI1
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat:37]: The most important thing to check is the pin assignments from Betaflight. If they have based this on the same board as you then maybe no changes are needed.

**Comment by andyp1per** [comment]
Make sure you are following https://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

You commits are in the subsystem "AP_HAL_ChibiOS"

File: libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat (hunk 4 above)
Comment (tenten8401) [reply libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat:37]: They should be correct according to the betaflight target, I will test everything on my board before changing from a draft

File: libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat (hunk 3 above)
Comment (tenten8401) [reply libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat:8]: Thanks, wasn't sure what I was supposed to do there. I'll see if I can find the list and change it.

File: libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat (hunk 2 above)
Comment (tenten8401) [reply libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat:75]: That fixed one of the compile issues, thanks for catching that. I'll see if I can find a definitive chart on which ones of these are actually USARTs, although most of the serial ports from the MatekF405 lined up with this pinout so I'm pretty sure they're correct. Is there a way to test which one is a USART?

Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/PyroF4OSD/hwdef.dat:75]: No idea :)

**Comment by tenten8401** [comment]
It's alive!!
Pic

I'll work on testing the board further and make sure everything works, getting some "Baro: unable to calibrate" critical errors, despite having `define HAL_BARO_ALLOW_INIT_NO_BARO` in there so I'll have to figure that out. Other than that initial boot is looking good.

**Comment by tenten8401** [comment]
I'm going to close this for now as my current priority is the Mamba F405 board and this one has a finnicky USB connector on it + no I2C ports. The baro issue mentioned above has been fixed.

**Pull request closed without merging**